### PR TITLE
Separate ARM vs data-plane poller files and generation

### PR DESCRIPTION
**Pull request by catalinaperalta**
This PR separates arm poller code generation from data-plane code generation, a slightly modified version of arm generation is used as a placeholder for dataplane generated code. 
Adding azartifacts to help track LRO progress for data-plane pollers.

### CODE REVIEW

File: src/generator/generator.ts [hunk 1]
```diff
@@ -59,12 +61,22 @@ export async function protocolGen(host: Host) {
     if (pagers.length > 0) {
       host.WriteFile(`${filePrefix}pagers.go`, pagers, undefined, 'source-file-go');
     }
-    const pollers = await generatePollers(session);
-    if (pollers.length > 0) {
-      const pollingHelper = await generatePollersHelper(session);
-      host.WriteFile(`${filePrefix}pollers_helper.go`, pollingHelper, undefined, 'source-file-go');
-      host.WriteFile(`${filePrefix}pollers.go`, pollers, undefined, 'source-file-go');
+    if (openapiType === 'data-plane') {
```
Comment (jhendrixMSFT) [inline src/generator/generator.ts:64]: Can this logic be pushed into `generatePollers()`, and have it return the correct content?

File: test/synapse/2019-06-01/azartifacts/pollers_helper.go [hunk 2]
```diff
@@ -0,0 +1,768 @@
...
+const (
+	operationInProgress string = "InProgress"
+	operationCanceled   string = "Canceled"
+	operationFailed     string = "Failed"
+	operationSucceeded  string = "Succeeded"
+)
+
+var pollingCodes = [...]int{http.StatusNoContent, http.StatusAccepted, http.StatusCreated, http.StatusOK}
+
+type pollingTracker interface {
```
Comment (jhendrixMSFT) [inline test/synapse/2019-06-01/azartifacts/pollers_helper.go:35]: This is all the ARM stuff.  Can we have this file generate a little as possible?  Once we start generating real content the diff will be extremely noisy.

File: src/generator/generator.ts (hunk 1 above)
Comment (catalinaperalta) [reply src/generator/generator.ts:64]: So I personally think it's better to keep arm poller code separate from data-plane poller code and would like to keep them in two separate places. I can add this logic in generatePollers and then call the corresponding func from there.

File: test/synapse/2019-06-01/azartifacts/pollers_helper.go (hunk 2 above)
Comment (catalinaperalta) [reply test/synapse/2019-06-01/azartifacts/pollers_helper.go]: done

File: src/generator/generator.ts (hunk 1 above)
Comment (catalinaperalta) [reply src/generator/generator.ts]: done

Comment (jhendrixMSFT) [reply src/generator/generator.ts]: Agreed we should keep them separate.  But generator.ts shouldn't need to know anything about ARM/data-plane, it just calls some generic `createPollers()` func and the correct content is returned.

File: test/synapse/2019-06-01/azartifacts/pollers.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	Done() bool
+	Poll(ctx context.Context) (*http.Response, error)
+	FinalResponse(ctx context.Context) (*CreateDataFlowDebugSessionResponseResponse, error)
+	ResumeToken() (string, error)
+}
+
+type createDataFlowDebugSessionResponsePoller struct {
+	// the client for making the request
+	pipeline azcore.Pipeline
+	pt       pollingTracker
```
Comment (jhendrixMSFT) [inline test/synapse/2019-06-01/azartifacts/pollers.go:30]: Where is pollingTracker defined?  Should it be deleted?

Comment (catalinaperalta) [reply test/synapse/2019-06-01/azartifacts/pollers.go]: Deleted. Done.

File: src/generator/arm_pollers.ts [hunk 4]
```diff
@@ -302,7 +302,7 @@ export async function generatePollers(session: Session<CodeModel>): Promise<stri
 
 
 // Creates the content in pollers_helper.go
-export async function generatePollersHelper(session: Session<CodeModel>): Promise<string> {
```
Comment (jhendrixMSFT) [inline src/generator/arm_pollers.ts:305]: Doesn't this go away with your recent changes to armcore?

Comment (catalinaperalta) [reply src/generator/arm_pollers.ts:305]: Yes, but that change will be implemented in a future PR. The code I have for the armcore implementation is ready to go once we get this in.

File: test/synapse/2019-06-01/azartifacts/notebook.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+// notebookOperations implements the NotebookOperations interface.
+type notebookOperations struct {
+	*Client
+}
+
+// CreateOrUpdateNotebook - Creates or updates a Note Book.
+func (client *notebookOperations) BeginCreateOrUpdateNotebook(ctx context.Context, notebookName string, notebook NotebookResource, notebookCreateOrUpdateNotebookOptions *NotebookCreateOrUpdateNotebookOptions) (*NotebookResourcePollerResponse, error) {
+	return nil, nil
```
Comment (catalinaperalta) [inline test/synapse/2019-06-01/azartifacts/notebook.go:43]: @jhendrixMSFT FYI, this is how I'm outputting the operations at the moment. Or should I add an errors.New("Not yet implemented"). Processing and initializing a request for a data-plane poller is still TBD and the helper code is not being outputted.

Comment (jhendrixMSFT) [reply test/synapse/2019-06-01/azartifacts/notebook.go:43]: Probably best to emit it as `errors.New("NYI")` just to be safe.

File: src/generator/models.ts [hunk 6]
```diff
@@ -553,7 +553,7 @@ function generateAliasType(structDef: StructDef, receiver: string, forMarshal: b
     if (prop.schema.type === SchemaType.DateTime) {
       text += `\t\t${prop.language.go!.name} *${prop.schema.language.go!.internalTimeType} \`${structDef.Language.marshallingFormat}:"${sn}"\`\n`;
     } else if (prop.language.go!.isAdditionalProperties || prop.language.go!.needsXMLDictionaryUnmarshalling) {
-      text += `\t\t${prop.language.go!.name} *additionalProperties \`${structDef.Language.marshallingFormat}:"${sn}"\`\n`;
+      text += `\t\t${prop.language.go!.name} *map[string]string \`${structDef.Language.marshallingFormat}:"${sn}"\`\n`;
```
Comment (jhendrixMSFT) [inline src/generator/models.ts]: This is incorrect.  The `additionalProperties` type is defined in a helper that's only emitted when `needsXMLDictionaryUnmarshalling` is true.  The fact that this is being emitted when the helper isn't suggests that the above condition has a bug.

Comment (jhendrixMSFT) [reply src/generator/models.ts]: It's also interesting that this failed only for Go 1.14.  Do you know why that is?

**Pull request merged**